### PR TITLE
Proving costs part 2

### PR DESCRIPTION
**Pull request by wackerow**
### `lib/proofs.ts`
- Updates all logic relating to proving costs to remove usage of `proving_cost` field, and instead calculate proving costs based on `proving_time` (in milliseconds) and the associated `hourly_rate` for the AWS `instance_type_id` that is provided for the proof's `cluster_id` in the `cluster_configurations` table (TS types updated—see below)
- Creates concept of "Stats", with four new helper function: `getCostPerProofStats`, `getCostPerMgasStats`, `getProvingTimeStats`, `getTotalTTPStats`
- These export an object with five fields to be consumed by the other front end components for each of these metrics:
  ```ts
  export type Stats = {
    avg: number
    avgFormatted: string
    best: number
    bestFormatted: string
    bestProof: Proof
  }
  ```

With the `get*Stats` helpers above, the no-longer-needed logic from the `columns.tsx` files were removed or simplified.

### Other `/lib` changed
- **`constants.ts`**: `AVERAGE_LABEL` added to have a canonical way we express the term "avg" (we had three forms before, just trying to remove ambiguity... can always override individual ones as-needed)
- **`number.ts`**: Added `formatUsd`
- **`time.ts`**: Crated; move `getTime` here, and added internal `prettyMs` utilizing `prettyMilliseconds` package
- **`types.ts`**: Updated naming so the primary named component is the union of the table type and extensions, ie. `export type Proof = ProofBase & ProofExtensions` pattern used before where `ProofBase` is the row type from the "proof" table. (Added intellisense comments)

Some other naming updates and adjustments now that more of the app has taken shape

## Preview link
https://deploy-preview-142--ethproofs.netlify.app/

### CODE REVIEW

File: app/page.tsx [hunk 1]
```diff
@@ -44,7 +47,15 @@ export default async function Index() {
...
-    .select("*,proofs!inner(id:proof_id,*,clusters(*))")
+    .select(
+      `*,proofs!inner(id:proof_id,*,
+        clusters(*,
+          cluster_configurations(*,
+            aws_instance_pricing(*)
+          )
+        )
+      )`
+    )
```
Comment (wackerow) [inline app/page.tsx:58]: <details><summary>Log output from Netlify on test branch</summary>

```
Dec 7, 10:29:37 AM: 33af605d INFO   *,proofs!inner(id:proof_id,*,
        clusters(*,
          cluster_configurations(*,
            aws_instance_pricing(*)
          )
        )
      )

Dec 7, 10:29:37 AM: 33af605d INFO   Logging `cluster_configurations` for proofs[0] of blocks[0]

Dec 7, 10:29:37 AM: 33af605d INFO   [
  {
    id: 1,
    cluster_id: 'b490d69e-2443-5d6d-bb00-3cdd8d5f181b',
    instance_count: 7,
    instance_type_id: 1,
    aws_instance_pricing: {
      id: 1,
      vcpu: 191,
      region: 'Idis iucunt everitur neque mucius solum am, quae neglexerat chaere nunc relique quae tenentem omnem.',
      created_at: '2024-12-05T15:51:31.112876+00:00',
      hourly_price: 0.0416,
      instance_type: 't3.medium',
      instance_memory: 12.8765,
      instance_storage: 'Sic ad locat quidem id, placeat loqueretur modus qui omni tam sublata principerit.'
    }
  },
  {
    id: 2,
    cluster_id: 'b490d69e-2443-5d6d-bb00-3cdd8d5f181b',
    instance_count: 2,
    instance_type_id: 2,
    aws_instance_pricing: {
      id: 2,
      vcpu: 24,
      region: 'Id sunt democculta ad naturque si quae.',
      created_at: '2024-12-05T15:51:31.112876+00:00',
      hourly_price: 0.0832,
      instance_type: 't3.large',
      instance_memory: 4.988,
      instance_storage: 'Cum tribus in pro ullam quid credo totelego, forte quod ne commundi omnis iisque ne probari.'
    }
  },
  {
    id: 4,
    cluster_id: 'b490d69e-2443-5d6d-bb00-3cdd8d5f181b',
    instance_count: 2,
    instance_type_id: 1,
    aws_instance_pricing: {
      id: 1,
      vcpu: 191,
      region: 'Idis iucunt everitur neque mucius solum am, quae neglexerat chaere nunc relique quae tenentem omnem.',
      created_at: '2024-12-05T15:51:31.112876+00:00',
      hourly_price: 0.0416,
      instance_type: 't3.medium',
      instance_memory: 12.8765,
      instance_storage: 'Sic ad locat quidem id, placeat loqueretur modus qui omni tam sublata principerit.'
    }
  },
  {
    id: 5,
    cluster_id: 'b490d69e-2443-5d6d-bb00-3cdd8d5f181b',
    instance_count: 3,
    instance_type_id: 1,
    aws_instance_pricing: {
      id: 1,
      vcpu: 191,
      region: 'Idis iucunt everitur neque mucius solum am, quae neglexerat chaere nunc relique quae tenentem omnem.',
      created_at: '2024-12-05T15:51:31.112876+00:00',
      hourly_price: 0.0416,
      instance_type: 't3.medium',
      instance_memory: 12.8765,
      instance_storage: 'Sic ad locat quidem id, placeat loqueretur modus qui omni tam sublata principerit.'
    }
  }
]
```

<p>



</p>
</details>

**Comment by wackerow** [comment]
Pending completion of migrations to update the cost calculations for the two existing views (teams_summary and recent_summary)

**Comment by wackerow** [comment]
@pettinarip Should be ready for a review... ideally anything from here is broken off as a separate task

File: supabase/migrations/20241210045518_update_pricing_logic.sql [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  join "public"."aws_instance_pricing" "a" on "cc"."instance_type_id" = "a"."id"
+  where "b"."timestamp" >= (now() - '30 days'::interval);
+
+drop view if exists "public"."teams_summary";
+
+create or replace view "public"."teams_summary" with (security_invoker=on) as
+  select "t"."team_id",
+         "t"."team_name",
+         "t"."logo_url",
+         coalesce(sum(cc."instance_count" * a."hourly_price" * (p."proving_time" / (1000.0 * 60 * 60))) / nullif(count(p."proof_id"), 0), 0) as "avg_cost_per_proof",
```
Comment (wackerow) [inline supabase/migrations/20241210045518_update_pricing_logic.sql:22]: Double check the summation here

Comment (pettinarip) [inline supabase/migrations/20241210045518_update_pricing_logic.sql:22]: yeah, I think we need to re-think this. We need to calc the `avg(cost_per_proof)`. So, we first need to calculate the individual cost for each proof and then, the avg of those per cluster per team..

I don't think we will be able to do this in just one select...will need to undercover old sql books xD

Comment (wackerow) [reply supabase/migrations/20241210045518_update_pricing_logic.sql:22]: @pettinarip

Okay, so we're trying to calculate the **Average Cost Per Proof**, which = **total costs** / **total number of proofs**

Below we have `and "p"."proof_status" = 'proved'` to filter out proofs not yet finished.

We're summing up the products of three things:
- cc."instance_count"
- a."hourly_price"
- And p."proving_time" (with ms/hours adjustment)

This gives us the **total costs**. We don't need to `sum(instance_count*hourly_price)` and _then_ multiply by `proving_time`, we can also just multiply each `instance_count*hourly_price` by `proving_time` and sum those... same mathematical result.

Then our **total number of proofs** is just `count(p."proof_id")`, but we use `nullif(count(p."proof_id"), 0)` to make sure we're not dividing by zero.

I think this works.

Comment (wackerow) [reply supabase/migrations/20241210045518_update_pricing_logic.sql:22]: Example:

```
1 * 5 = 5
2 * 4 = 8
3 * 3 = 9
---------
       22 * 10 = 220

```


```
1 * 5 * 10 = 50
2 * 4 * 10 = 80
3 * 3 * 10 = 90
---------------
            220
```

Comment (pettinarip) [reply supabase/migrations/20241210045518_update_pricing_logic.sql:22]: yeah, I think this looks fine! I got confused on the how we were getting the cost per machine.

**Review by pettinarip — APPROVED** [review]
Awesome @wackerow I think this looks fine. Removing one of the migration files and merging

File: supabase/migrations/20241208063655_update_team_summary_view.sql
[the remark this answers is not in the document]
Comment (wackerow) [reply supabase/migrations/20241208063655_update_team_summary_view.sql:1]: Yes, thanks for committing that! I had forgotten 🙏

File: supabase/migrations/20241210045518_update_pricing_logic.sql (hunk 2 above)
Comment (wackerow) [reply supabase/migrations/20241210045518_update_pricing_logic.sql:22]: Comments below, should be good

**Pull request merged**